lsp/nls/src/requests/completion.rs

```diff
@@ -1,16 +1,99 @@
 use codespan::ByteIndex;
 use codespan_lsp::position_to_byte_index;
 use log::debug;
-use lsp_server::{RequestId, Response, ResponseError};
+use lsp_server::{ErrorCode, RequestId, Response, ResponseError};
 use lsp_types::{CompletionItem, CompletionParams};
+use nickel_lang::{
+    identifier::Ident,
+    term::{Contract, MetaValue},
+    types::{AbsType, RowIteratorItem, Types},
+};
 use serde_json::Value;
 
 use crate::{
-    linearization::interface::TermKind,
+    linearization::{
+        interface::{TermKind, UsageState, ValueState},
+        LinearizationItem,
+    },
     server::Server,
     trace::{Enrich, Trace},
 };
 
+/// Find record fields for an item with the specified id.
+fn find_record_fields(
+    linearization: &Vec<LinearizationItem<Types>>,
+    id: usize,
+) -> Option<Vec<Ident>> {
+    linearization.iter().find_map(|item| match item.kind {
+        TermKind::Record(ref fields) if item.id == id => {
+            Some(fields.keys().map(|ident| ident.clone()).collect())
+        }
+        TermKind::Declaration(_, _, ValueState::Known(new_id))
+        | TermKind::Usage(UsageState::Resolved(new_id))
+            if item.id == id =>
+        {
+            find_record_fields(linearization, new_id)
+        }
+        _ => None,
+    })
+}
+
+/// Find a record contract of the item with the specified id.
+fn find_record_contract(
+    linearization: &Vec<LinearizationItem<Types>>,
+    id: usize,
+) -> Option<Contract> {
+    linearization.iter().find_map(|item| match &item.meta {
+        Some(MetaValue { contracts, .. }) if item.id == id => {
+            contracts.iter().find_map(|contract| {
+                if let Types(AbsType::Record(..)) = contract.types {
+                    Some(contract.clone())
+                } else {
+                    None
+                }
+            })
+        }
+        _ if item.id == id => match item.kind {
+            TermKind::Declaration(_, _, ValueState::Known(new_id))
+            | TermKind::Usage(UsageState::Resolved(new_id)) => {
+                find_record_contract(&linearization, new_id)
+            }
+            _ => None,
+        },
+        _ => None,
+    })
+}
+
+/// Extract identifiers from a row type which forms a record.
+/// Ensure that ty is really a row type.
+fn extract_ident(ty: &Box<Types>) -> Vec<Ident> {
+    ty.iter_as_rows()
+        .filter_map(|item| match item {
+            RowIteratorItem::Row(ident, _) => Some(ident.clone()),
+            RowIteratorItem::Tail(..) => None,
+        })
+        .collect::<Vec<_>>()
+}
+
+/// Get the identifier before the `.`, for record completion.
+fn get_identifier(text: &str) -> Result<String, ResponseError> {
+    let name: String = text
+        .chars()
+        .rev()
+        .skip_while(|c| *c == '.') // skip . (if any)
+        .collect();
+
+    // unwrap is safe here because we know this is a correct regex.
+    let regex = regex::Regex::new(r"[_a-zA-Z0-9-']*[a-zA-Z]_?").unwrap();
```

Comment:
I think this regexp is wrong: there are many other characters allowed in identifiers in Rust, according to https://doc.rust-lang.org/reference/identifiers.html
Counter-examples:
* `京`
* `к`

E.g. I just tested with `fooкbar.` and it returns `bar` rather than `fooкbar`.

Tests, in particular PBTs, would be quite helpful in checking this function is well-behaved

Reply:
Well, this is the reverse (because we reversed the source text) of what is in the Nickel lexer, so I think it should be good for Nickel source files: 
```rust
 #[regex("_?[a-zA-Z][_a-zA-Z0-9-']*")]
    Identifier(&'input str),
```

Reply:
I don't think Nickel supports unicode characters in identifiers

Reply:
How about linking to https://nickel-lang.org/user-manual/syntax in a comment here for reference? To let the reader know exactly what characters form valid identifiers